e2e/test/AppContextTests.cs

```diff
@@ -0,0 +1,332 @@
+#endif
+
+            if (isFlagSet)
+            {
+                await ReceiveAsyncAfterDisposeAsync(TestDeviceType.Sasl, Client.TransportType.Amqp_Tcp_Only, true).ConfigureAwait(false);
+            }
+            else
+            {
+                Assert.Fail("AppContext flag DisableObjectDisposedExceptionForReceiveAsync not set");
+            }
+        }
+
+        [Ignore("ReceiveAsync hangs due to bug in DotNetty (https://github.com/Azure/DotNetty/issues/502)")]
+        [TestMethod]
+        public async Task AppConfig_ReceiveAsyncAfterDispose_Sasl_Mqtt()
+        {
+
+            bool isFlagSet = false;
+            bool flag = false;
+
+#if !NET451
+            isFlagSet = AppContext.TryGetSwitch(AppContextConstants.DisableObjectDisposedExceptionForReceiveAsync, out flag);
+#endif
+
+            if (isFlagSet)
+            {
+                await ReceiveAsyncAfterDisposeAsync(TestDeviceType.Sasl, Client.TransportType.Mqtt_Tcp_Only, false).ConfigureAwait(false);
+            }
+            else
+            {
+                Assert.Fail("AppContext flag DisableObjectDisposedExceptionForReceiveAsync not set");
+            }
+        }
+
+        [Ignore("ReceiveAsync hangs due to bug in DotNetty (https://github.com/Azure/DotNetty/issues/502)")]
+        [TestMethod]
+        public async Task AppConfig_ReceiveAsyncWithCancellationTokenAfterDispose_Sasl_Mqtt()
+        {
+
+            bool isFlagSet = false;
+            bool flag = false;
+
+#if !NET451
+            isFlagSet = AppContext.TryGetSwitch(AppContextConstants.DisableObjectDisposedExceptionForReceiveAsync, out flag);
+#endif
+
+            if (isFlagSet)
+            {
+                await ReceiveAsyncAfterDisposeAsync(TestDeviceType.Sasl, Client.TransportType.Mqtt_Tcp_Only, true).ConfigureAwait(false);
+            }
+            else
+            {
+                Assert.Fail("AppContext flag DisableObjectDisposedExceptionForReceiveAsync not set");
+            }
+        }
+
+        private async Task ReceiveAsyncAfterDisposeAsync(TestDeviceType testDeviceType, Client.TransportType transportType, bool expectException)
+        {
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix, testDeviceType).ConfigureAwait(false);
+
+            using (CancellationTokenSource cts = new CancellationTokenSource(TimeSpan.FromSeconds(60)))
+            using (DeviceClient deviceClient = testDevice.CreateDeviceClient(transportType))
+            {
+                Exception exceptionCaught = null;
+                Client.Message message = null;
+
+                try
+                {
+                    await deviceClient.OpenAsync().ConfigureAwait(false);
+
+                    Task<Client.Message> t;
+
+                    if (expectException)
+                    {
+                        // ReceiveAsync with cancellation token should always throw exception.
+                        t = deviceClient.ReceiveAsync(cts.Token);
```

Comment:
I don't understand this block, which scenario does this include?
Isn't ReceiveAsync with cancellation token supported for DeviceClient: https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/device/src/DeviceClient.cs#L288?

Reply:
Is this change only relevant when we pass a cancellation token to the ReceiveAsync API? What is the behavior for ReceiveAsync() and ReceiveAsync(Timespan timeout)?